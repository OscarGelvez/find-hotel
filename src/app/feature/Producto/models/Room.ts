export interface IRoom {
  /**
   * Identificador del hotel
   */
  id: number;
  /**
   * Objeto hotel al cual pertenece la habitación
   */
  hotel: IHotel;
  /**
   * Valor de la habitación
   */
  value: number;
  /**
   * Indica si tiene o no aire acondicionado
   */
  ac: string;
  /**
   * Indica si cuenta o no con parquedadero
   */
  parking: string;
  /**
   * Fecha inicial de disponibilidad
   */
  available_from: string;
  /**
   * Fecha final de disponibilidad
   */
  available_until: string;
  /**
   * Indica si cuenta con conexión wifi
   */
  wifi: string;
  /**
   * Capacidad de personas por habitación
   */
  capacity: number;
  /**
   * Descripción general de la habitación
   */
  description: string;
}

export interface IHotel {
  /**
   * Identificador del hotel
   */
  id: number;
  /**
   * Nombre del hotel
   */
  name: string;
  /**
   * Dirección y/o ubicacion del hotel
   */
  address: string;
  /**
   * Cantidad de estrellas del hotel
   */
  stars: number;
  /**
   * Descripción general
   */
  description: string;
  /**
   * Puntaje de calificación dado por usuarios
   */
  score: number;
}
