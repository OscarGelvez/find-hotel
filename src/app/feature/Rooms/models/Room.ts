export interface IRoom {
  /**
   * Identificador del hotel
   */
  id: number;
  /**
   * Título de la habitación
   */
  title: string;
  /**
   * Imagen descriptiva de la habitación
   */
  image: string;
  /**
   * Objeto hotel al cual pertenece la habitación
   */
  hotel: IHotel;
  /**
   * Valor de la habitación
   */
  value: string;
  /**
   * Indica si tiene o no aire acondicionado
   */
  ac: string;
  /**
   * Indica si cuenta o no con parquedadero
   */
  parking: string;
  /**
   * Fecha inicial de disponibilidad
   */
  available_from: string;
  /**
   * Fecha final de disponibilidad
   */
  available_until: string;
  /**
   * Indica si cuenta con conexión wifi
   */
  wifi: string;
  /**
   * Capacidad de personas por habitación
   */
  capacity: number;
  /**
   * Descripción general de la habitación
   */
  description: string;
}

export interface IHotel {
  /**
   * Identificador del hotel
   */
  id: number;
  /**
   * Nombre del hotel
   */
  name: string;
  /**
   * Dirección y/o ubicacion del hotel
   */
  address: string;
  /**
   * Cantidad de estrellas del hotel
   */
  stars: number;
  /**
   * Descripción general
   */
  description: string;
  /**
   * Puntaje de calificación dado por usuarios
   */
  score: number;
}

export interface IFieldsForm {
  /**
   * Fecha inicio de reservación
   */
  fromDate?: string;
  /**
   * Fecha final de reservación
   */
  untilDate?: string;
  /**
   * Cantidad de personas en habitación
   */
  capacity?: string;
  /**
   * Habitación con aire acondicionado
   */
  withAir?: boolean;
  /**
   * Habitación con parqueadero propio
   */
  withParking?: boolean;
}

export interface IFieldsFormBooking {
  /**
   * Nombre de ususario que hace la reserva
   */
  name: string;
  /**
   * Correo electrónico de quien hacer la reserva
   */
  email: string;
  /**
   * Tipo de identificación
   */
  identification_type: string;
  /**
   * Número de identificación
   */
  identification: string;
}

export interface IBookingData {
  /**
   * Objeto con información de la habitación seleccionada
   */
  roomData: IRoom;
  /**
   * Objeto con la información el usuari oque reserva
   */
  bookData: IFieldsFormBooking;
}

export interface IFieldsFormFindBooking {
  /**
   * Correo electrónico de quien hacer la reserva
   */
  email: string;
  /**
   * Tipo de identificación
   */
  identification_type: string;
  /**
   * Número de identificación
   */
  identification: string;
}
