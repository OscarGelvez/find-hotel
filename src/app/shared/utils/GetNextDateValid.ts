/**
 * Valida la siguiente fecha valida del campo Fecha de salida
 * según la fecha seleccionada en el campo Fecha ingreso
 * @param currentDate Fecha actual
 * @param fromDate Fecha inicial
 * @param minDate mínima fecha incial
 * @returns
 */
export const getNextDate = (
  currentDate: Date,
  fromDate: string,
  minDate: string
) => {
  if (fromDate === minDate) {
    currentDate.setDate(currentDate.getDate() + 1);
    return currentDate.toISOString().split('T')[0];
  } else {
    const auxDate = fromDate !== '' ? new Date(fromDate) : currentDate;
    auxDate.setDate(auxDate.getDate() + 1);
    return auxDate.toISOString().split('T')[0];
  }
};
