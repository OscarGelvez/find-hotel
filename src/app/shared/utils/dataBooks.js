export const bookInfo = {
  bookData: {
    name: 'oscar andres',
    email: 'oscargelvez23@gmail.com',
    identification_type: '1',
    identification: '1323131232',
  },
  roomData: {
    id: 54,
    title: 'Peace and Love',
    image:
      'https://dynamic-media-cdn.tripadvisor.com/media/photo-o/0d/2d/70/6d/habitacion-master-suite.jpg?w=600&h=300&s=1',
    hotel: {
      id: 3,
      name: 'Cencor hotel',
      address: 'Calle 1 # 16-23',
      stars: 4,
      description: 'Ambiente familiar, amigos, descaso y relajación',
      score: 7.2,
    },
    value: '56.000',
    ac: 'N',
    parking: 'N',
    available_from: '2021-05-02',
    available_until: '2021-05-04',
    wifi: 'N',
    capacity: 1,
    description: 'Baño privado, cama sencilla, TV.',
  },
  id: 5,
};
