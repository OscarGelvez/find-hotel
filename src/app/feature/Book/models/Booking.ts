import { IRoom } from 'app/feature/Rooms/models/Room';

export interface IFieldsFormBooking {
  /**
   * Nombre de ususario que hace la reserva
   */
  name: string;
  /**
   * Correo electrónico de quien hacer la reserva
   */
  email: string;
  /**
   * Tipo de identificación
   */
  identification_type: string;
  /**
   * Número de identificación
   */
  identification: string;
}

export interface IBooking {
  /**
   * Objeto con información de la habitación seleccionada
   */
  roomData: IRoom;
  /**
   * Objeto con la información el usuari oque reserva
   */
  bookData: IFieldsFormBooking;
  /**
   * Identificador en base de datos
   */
  id?: number;
}

export interface IFieldsFormFindBooking {
  /**
   * Correo electrónico de quien hacer la reserva
   */
  email: string;
  /**
   * Tipo de identificación
   */
  identification_type: string;
  /**
   * Número de identificación
   */
  identification: string;
}
