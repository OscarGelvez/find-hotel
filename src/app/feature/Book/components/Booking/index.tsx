import { fieldsFormFindBooking } from 'app/feature/Rooms/models/Room';
import { Button } from 'app/shared/components/Button';
import { Field, Form, Formik, FormikHelpers } from 'formik';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as Yup from 'yup';

interface BookingDetail {
  onFindBooking: (findBookingData: fieldsFormFindBooking) => void;
}

const initialValues = {
  email: '',
  identification_type: '',
  identification: '',
};

const validationSchema = Yup.object().shape<fieldsFormFindBooking>({
  email: Yup.string().email('Email no válido').required('Necesitamos tu email'),
  identification_type: Yup.string().required(
    'Necesitamos tu tipo de identificación'
  ),
  identification: Yup.string().required('Necesitamos tu identificación'),
});

export const BookingDetail: React.FC<BookingDetail> = ({ onFindBooking }) => {
  const handleSubmit = (
    values: fieldsFormFindBooking,
    { resetForm }: FormikHelpers<fieldsFormFindBooking>
  ) => {
    onFindBooking({
      email: values.email,
      identification_type: values.identification_type,
      identification: values.identification,
    });
    resetForm();
  };
  return (
    <>
      <div data-testid="find-rooms" className="card-rooms">
        <div className="row">
          <div className="card filter-area ">
            <div className="col">
              <Formik
                initialValues={initialValues}
                validationSchema={validationSchema}
                onSubmit={handleSubmit}
              >
                {({ errors, touched }) => (
                  <Form data-testid="form-find-rooms">
                    <div className="form-group row mb-3 mt-3">
                      <div className=" col-12 col-md-4">
                        <label
                          htmlFor="email"
                          className="col-12 col-form-label"
                        >
                          Correo electrónico
                        </label>
                        <Field
                          name="email"
                          type="email"
                          className="form-control col-12"
                          placeholder="eje: joe@email.com"
                        />
                        {errors.email && touched.email ? (
                          <small className="text-danger">{errors.email}</small>
                        ) : null}
                      </div>
                      <div className="col-12 col-md-4">
                        <label className="pr-2" htmlFor="identification_type">
                          Tipo de identificación
                        </label>
                        <Field
                          as="select"
                          className=" d-block w-100"
                          aria-label=".form-select-sm example"
                          id="identification_type"
                          name="identification_type"
                        >
                          <option value="">Seleccione</option>
                          <option value="1">Cédula de ciudadanía</option>
                          <option value="2">Cédula de extranjeria</option>
                          <option value="3">Nit</option>
                          <option value="4">Pasaporte</option>
                        </Field>
                        {errors.identification_type &&
                        touched.identification_type ? (
                          <small className="text-danger">
                            {errors.identification_type}
                          </small>
                        ) : null}
                      </div>
                      <div className=" col-12 col-md-4">
                        <label className="pr-2" htmlFor="identification">
                          Identificación
                        </label>
                        <Field
                          name="identification"
                          className="form-control"
                          placeholder="eje: 60830147"
                        />
                        {errors.identification && touched.identification ? (
                          <small className="text-danger">
                            {errors.identification}
                          </small>
                        ) : null}
                      </div>
                    </div>
                    <div className="row text-center">
                      <Button type="submit" className="btn ">
                        Buscar
                      </Button>
                    </div>
                  </Form>
                )}
              </Formik>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

BookingDetail.propTypes = {
  // msg: PropTypes.string.isRequired,
};
