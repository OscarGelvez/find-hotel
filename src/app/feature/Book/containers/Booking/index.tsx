import {
  IBooking,
  IFieldsFormFindBooking,
} from 'app/feature/Book/models/Booking';
import * as PropTypes from 'prop-types';
import * as React from 'react';

import { FindBookingForm } from '../../components/findBooking';
import { ListBookings } from '../../components/listBooking';
import { DivContainer } from './styles';

interface IBookingsProps {
  findBooking: (findBookingData: IFieldsFormFindBooking) => void;
  listBookingsFinded: Array<IBooking>;
  cancelBooking: (bookId: number) => void;
  isLoading: boolean;
}

export const Booking: React.FC<IBookingsProps> = ({
  findBooking,
  listBookingsFinded,
  cancelBooking,
  isLoading,
}) => {
  return (
    <DivContainer data-testid="book">
      <FindBookingForm onFindBooking={findBooking} />
      <ListBookings
        listBookingsFinded={listBookingsFinded}
        cancelBooking={cancelBooking}
        isLoading={isLoading}
      />
    </DivContainer>
  );
};

Booking.propTypes = {
  findBooking: PropTypes.func.isRequired,
  listBookingsFinded: PropTypes.array.isRequired,
  cancelBooking: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
};
