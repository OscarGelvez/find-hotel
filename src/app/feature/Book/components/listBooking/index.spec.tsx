import '@testing-library/jest-dom';

import { cleanup, render } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { bookInfo } from '../../../../shared/utils/dataBooks';
import { IBooking } from '../../models/Booking';
import { ListBookings } from './index';

const newBook: IBooking = bookInfo;
afterEach(cleanup);

it('rederizar el componente ListBookings sin errores', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <ListBookings listBookingsFinded={[]} cancelBooking={() => null} />,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

it('renderiza el componente ListBookings y verifica existencia de componentes hijos', () => {
  const { getByTestId } = render(
    <ListBookings listBookingsFinded={[newBook]} cancelBooking={() => null} />
  );

  expect(getByTestId('list-bookings')).toContainElement(
    getByTestId('modal-detail-book')
  );
  expect(getByTestId('list-bookings')).toContainElement(
    getByTestId('list-book-btn-detail')
  );
  expect(getByTestId('list-bookings')).toContainElement(
    getByTestId('list-book-btn-cancel')
  );
  expect(getByTestId('list-bookings')).toContainElement(
    getByTestId('modal-book-delete')
  );
});

it('Compara snapshot del componente ListBookings cuando hay reservas', () => {
  const element = renderer
    .create(
      <ListBookings listBookingsFinded={[newBook]} cancelBooking={() => null} />
    )
    .toJSON();
  expect(element).toMatchSnapshot();
});

it('Compara snapshot del componente ListBookings cuando no hay reservas', () => {
  const element = renderer
    .create(<ListBookings listBookingsFinded={[]} cancelBooking={() => null} />)
    .toJSON();
  expect(element).toMatchSnapshot();
});
