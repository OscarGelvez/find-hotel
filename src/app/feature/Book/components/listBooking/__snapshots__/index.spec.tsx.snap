// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compara snapshot del componente ListBookings cuando hay reservas 1`] = `
<div
  className="card-rooms"
  data-testid="list-bookings"
>
  <div
    className="row pt-5"
  >
    <div
      className="col"
    >
      <div
        className="card room-item"
        data-testid="card-information-bookigns"
      >
        <div
          className="card-header book-item-header "
        >
          <h3>
            Cencor hotel
          </h3>
        </div>
        <div
          className="card-body"
        >
          <div
            className="row"
          >
            <div
              className="col-12 col-md-6 col-lg-3"
            >
              <h5>
                A nombre de:
              </h5>
              <span>
                oscar andres
              </span>
              <h5
                className="mt-2"
              >
                Ingreso:
              </h5>
              <span>
                2021-05-02
              </span>
            </div>
            <div
              className="col-12 col-md-6 col-lg-3"
            >
              <h5>
                Habitación:
              </h5>
              <span>
                Peace and Love
              </span>
              <h5
                className="mt-2"
              >
                Salida:
              </h5>
              <span>
                2021-05-04
              </span>
            </div>
            <div
              className="col-12 col-md-6 col-lg-3"
            >
              <h5>
                Valor:
              </h5>
              <span>
                56.000
              </span>
              <h5
                className="mt-2"
              >
                Camas:
              </h5>
              <span>
                1
              </span>
            </div>
            <div
              className="col-12 col-md-6 col-lg-3"
            >
              <div
                aria-labelledby="modalDetailBookLabel"
                className="modal fade"
                data-testid="modal-detail-book"
                id="modalDetailBook"
              >
                <div
                  className="modal-dialog modal-xl"
                >
                  <div
                    className="modal-content"
                  >
                    <div
                      className="modal-header"
                    >
                      <h5
                        className="modal-title"
                        id="modalDetailBookLabel"
                      >
                        Detalles de tu reserva
                      </h5>
                      <button
                        aria-label="Close"
                        className="btn-close"
                        data-bs-dismiss="modal"
                        type="button"
                      />
                    </div>
                    <div
                      className="modal-body"
                    />
                  </div>
                </div>
              </div>
              <button
                className="btn btn-primary btn-book w-100"
                data-bs-target="#modalDetailBook"
                data-bs-toggle="modal"
                data-testid="list-book-btn-detail"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fas fa-check ml-3"
                />
                 Detalles
              </button>
              <button
                className="btn btn-danger btn-book w-100 mt-3"
                data-bs-target="#modalDeleteBook"
                data-bs-toggle="modal"
                data-testid="list-book-btn-cancel"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fas fa-times ml-3"
                />
                 Cancelar
              </button>
              <div
                aria-labelledby="modalDeleteBookLabel"
                className="modal fade"
                data-testid="modal-book-delete"
                id="modalDeleteBook"
              >
                <div
                  className="modal-dialog modal-xl"
                >
                  <div
                    className="modal-content"
                  >
                    <div
                      className="modal-header"
                    >
                      <h5
                        className="modal-title"
                        id="modalDeleteBookLabel"
                      >
                        Eliminar reserva
                      </h5>
                      <button
                        aria-label="Close"
                        className="btn-close"
                        data-bs-dismiss="modal"
                        type="button"
                      />
                    </div>
                    <div
                      className="modal-body"
                    />
                    <div
                      className="modal-footer"
                    >
                      <div
                        className="d-flex justify-content-between w-100"
                      >
                        <button
                          className="btn btn-secondary w-25"
                          data-bs-dismiss="modal"
                          type="button"
                        >
                          No
                        </button>
                        <button
                          className="btn btn-danger w-25"
                          onClick={[Function]}
                          type="button"
                        >
                          Sí, eliminar
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Compara snapshot del componente ListBookings cuando no hay reservas 1`] = `
<div
  className="card-rooms"
  data-testid="list-bookings"
>
  <div
    className="row pt-5"
  >
    <div
      className="col"
    >
      <div
        className="text-center"
      >
        <h5>
          No hay reservas para mostrar
        </h5>
      </div>
    </div>
  </div>
</div>
`;
